### PR TITLE
Cargo bounty cube incentives

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request

1. Bounty cube announces its creation over Supply with its creator and area, starting the speedy delivery bonus countdown
2. 20% bonus for cargo if you send the cube within 5 minutes of its creation
3. Using an export scanner on it in the supply shuttle awards a 10% personal handling tip when it reaches Centcom
4. Cube nags at five minutes over Supply if it's not sent and cargo bonus is lost
5. If there's a handler attached when it nags their ID card tells them it hasn't been sent
6. Cube continues to nag less and less often
7. All nags tell you the cube's area

## Why It's Good For The Game

Incentivises cargo to send bounty cubes in a timely fashion.

## Changelog
:cl: cacogen
add: Bounty cube announces creation over Supply with its creator and area
add: Bounty cube speedy delivery bonus of 20% for cargo if it's sent within five minutes
add: Bounty cube personal handling tip of 10% if you scan it with an export scanner while on the cargo shuttle (paid on sale)
add: Bounty cubes nag over Supply with their area, starting at five minutes with time increasing between each
/:cl:

### CODE REVIEW

**Comment by ArcaneMusic** [comment]
You've given bounty cubes tracking labels.
I'll admit, it's clever, I like this 👍

File: code/game/machinery/civilian_bounties.dm [hunk 1]
```diff
@@ -247,10 +244,75 @@
...
+	radio = new(src)
+	radio.keyslot = new radio_key
+	radio.listening = FALSE
+	radio.recalculateChannels()
+	START_PROCESSING(SSobj, src)
+	next_nag_time = world.time + nag_cooldown //grace period before the nagging begins
+
+/obj/item/bounty_cube/Destroy()
+	. = ..()
+	QDEL_NULL(radio)
```
Comment (coiax) [inline code/game/machinery/civilian_bounties.dm:285]: Do cleanup logic specific to the type before the `. = ..()` call.

File: code/game/machinery/civilian_bounties.dm [hunk 2]
```diff
@@ -247,10 +244,75 @@
...
+	///Percentage of the bounty payout received by the person who claims the handling tip.
+	var/handler_cut = 10
+	///Countdown timer that is used to indicate when cargo loses its speedy delivery bonus.
+	var/bonus_countdown
+	///Base time to nag Supply listeners to send this cube.
+	var/nag_cooldown = 5 MINUTES
+	///How much the time between nags extends each nag.
+	var/nag_cooldown_multiplier = 1.25
+	///Next world tick to nag Supply listeners.
+	var/next_nag_time = 0
```
Comment (coiax) [inline code/game/machinery/civilian_bounties.dm:260]: Use the COOLDOWN_DECLARE family of define helpers.

File: code/datums/components/pricetag.dm [hunk 3]
```diff
@@ -1,13 +1,16 @@
 /datum/component/pricetag
-	var/datum/bank_account/owner
-	var/profit_ratio = 1
+	var/default_profit_ratio = 1
+	var/list/payees = list()
```
Comment (coiax) [inline code/datums/components/pricetag.dm:3]: Could you put an autodoc string `///` above this explaining what it's likely to contain

File: code/modules/cargo/export_scanner.dm [hunk 4]
```diff
@@ -9,13 +9,30 @@
 	item_flags = NOBLUDGEON
 	w_class = WEIGHT_CLASS_SMALL
 
-/obj/item/export_scanner/afterattack(obj/O, mob/user, proximity)
+/obj/item/export_scanner/afterattack(obj/O, mob/living/carbon/user, proximity)
```
Comment (coiax) [inline code/modules/cargo/export_scanner.dm:12]: Is this an assertion you can make with `afterattack`?

File: code/game/machinery/civilian_bounties.dm [hunk 5]
```diff
@@ -247,10 +244,75 @@
...
+/obj/item/bounty_cube/Destroy()
+	. = ..()
+	QDEL_NULL(radio)
+
+/obj/item/bounty_cube/examine()
+	. = ..()
+	if(speed_bonus)
+		. += "<span class='notice'><b>[time2text(bonus_countdown,"mm:ss")]</b> remaining until <b>[bounty_value * (speed_bonus/100)] cr</b> speedy delivery bonus lost.</span>"
+	if(handler_cut && !bounty_handler_account)
+		. += "<span class='notice'>Scan this in the cargo shuttle with an export scanner to claim the <b>[bounty_value * (handler_cut/100)] cr</b> handling tip."
```
Comment (coiax) [inline code/game/machinery/civilian_bounties.dm:292]: Need to close this `<span>`

**Review by coiax — CHANGES_REQUESTED** [review]
An interesting idea, would like to make some changes to make things cleaner and more consistent.

My main concern is that this is going to get too spammy far too quickly. I admit, having 3 bounty cubes pile up is a concern, but that means that in the space of 12 minutes, Cargo is going to get 6 seperate messages about it.

**Comment by cacogen** [comment]
I understand. The potential spamminess is one of the concerns I have. But I'd prefer to see it in action before making assumptions about it. If it gets pulled I'll tweak the problems I see with it.

I don't know if six messages for three cubes is going to be all that much in twelve minutes of chat logs. If you send it in five minutes (and that seems reasonable barring someone taking off with it, which I don't think will happen often) then it's only the creation messages which are desirable because they tell you how to get your bonus. 

If it proves to be too spammy I'll have to figure out another way of getting the message to people with shuttle access or an alternative to using a message (which I assume would be more easily missed).

I was thinking of a console or a tablet app that has a list of cubes and their locations, but by itself that doesn't solve the problems of them being easy to ignore, undesirable to send and there being no motivation to send them quickly. The tablet app could let you track cubes using an overlay too.

**Comment by cacogen** [comment]
Ok so some issues

1. @coiax COOLDOWN_DECLARE seems to create a global variable and I'm not sure how to make it use a different variable for each cube
2. Value isn't split between cargo and the bounty holder, Centcom pays 130% of the value, should probably fix but this will devalue them for cargo (i.e. 70% vs. 100%, with 20% speed bonus optional in each case)
3. Cubes are trivially undervalued for reasons I don't understand (something to do with the k-elasticity I'm guessing), would like to fix for perfectionism but not sure how

Bonus

  i. Not sure if it can register multiple handlers but if so I need to fix that and it should probably not be overwriteable (though this creates drama so perhaps it should)
  ii. It sends the handler registered message after the handler has already been registered (makes it less clear if it worked so I will fix)

**Comment by cacogen** [comment]
I want to use the creation time variable I added here to tell the bounty holder how long the cube took to send, in addition to the name of the handler and the amount the holder got from it. Currently it's not used. Price tags can probably have custom sale messages set by their parent.

File: code/game/machinery/civilian_bounties.dm [hunk 6]
```diff
@@ -247,10 +244,100 @@
...
+		bonus_countdown = next_nag_time - world.time
+	if(COOLDOWN_FINISHED(src, next_nag_time))
+		if(!is_centcom_level(z) && !is_reserved_level(z)) //don't send message if we're on Centcom or in transit
+			radio.talk_into(src, "Unsent in <b>[get_area(src)]</b>.[speed_bonus ? " Speedy delivery bonus of <b>[bounty_value * speed_bonus]</b> credit\s lost." : ""]", RADIO_CHANNEL_SUPPLY)
+			speed_bonus = 0
+			bounty_holder_account.bank_card_talk("\The [src] is unsent in <b>[get_area(src)]</b>.")
+			if(bounty_handler_account)
+				bounty_handler_account.bank_card_talk("\The [src] is unsent in <b>[get_area(src)]</b>.")
+		nag_cooldown = nag_cooldown * nag_cooldown_multiplier
+		COOLDOWN_START(src, next_nag_time, world.time + nag_cooldown)
```
Comment (coiax) [inline code/game/machinery/civilian_bounties.dm:307]: COOLDOWN_START already uses `world.time`.
```suggestion
		COOLDOWN_START(src, next_nag_time, nag_cooldown)
```

File: code/game/machinery/civilian_bounties.dm [hunk 7]
```diff
@@ -247,10 +244,100 @@
...
 	///Value of the bounty that this bounty cube sells for.
 	var/bounty_value = 0
+	///Multiplier for the bounty payout received by the Supply budget if the cube is sent without having to nag.
+	var/speed_bonus = 0.2
+	///Multiplier for the bounty payout received by the person who completed the bounty.
+	var/holder_cut = 0.3
+	///Multiplier for the bounty payout received by the person who claims the handling tip.
+	var/handler_tip = 0.1
+	///Time until speed bonus expires.
+	var/bonus_countdown = 5 MINUTES
```
Comment (coiax) [inline code/game/machinery/civilian_bounties.dm:254]: I'm confused why this is a var, given that the speed bonus is lost when the cube first nags.

**Review by coiax — CHANGES_REQUESTED** [review]

File: code/game/machinery/civilian_bounties.dm [hunk 8]
```diff
@@ -247,10 +244,100 @@
...
+		. += "<span class='notice'><b>[time2text(bonus_countdown,"mm:ss")]</b> remains until <b>[bounty_value * speed_bonus]</b> credit speedy delivery bonus lost.</span>"
+	if(handler_tip && !bounty_handler_account)
+		. += "<span class='notice'>Scan this in the cargo shuttle with an export scanner to register your bank account for the <b>[bounty_value * handler_tip]</b> credit handling tip.</span>"
+
+/obj/item/bounty_cube/process(delta_time)
+	if(speed_bonus)
+		bonus_countdown = next_nag_time - world.time
+	if(COOLDOWN_FINISHED(src, next_nag_time))
+		if(!is_centcom_level(z) && !is_reserved_level(z)) //don't send message if we're on Centcom or in transit
+			radio.talk_into(src, "Unsent in <b>[get_area(src)]</b>.[speed_bonus ? " Speedy delivery bonus of <b>[bounty_value * speed_bonus]</b> credit\s lost." : ""]", RADIO_CHANNEL_SUPPLY)
```
[the remark this answers is not in the document]
Comment (cacogen) [reply code/game/machinery/civilian_bounties.dm:301]: Well, it obviously indicates emphasis in speech. Players not having that level of control doesn't mean other things that talk shouldn't. I doubt it's unprecedented in the codebase.

But this minor type of thing isn't worth arguing over. And the boldness is difficult to see in the monospaced font machines use anyway.

File: code/game/machinery/civilian_bounties.dm [hunk 9]
```diff
@@ -247,10 +244,100 @@
...
+	name = "\improper [bounty_value] cr [name]"
+	desc += " The sales tag indicates it was <i>[bounty_holder] ([bounty_holder_job])</i>'s reward for completing the <i>[bounty_name]</i> bounty."
+	AddComponent(/datum/component/pricetag, holder_id.registered_account, holder_cut)
+	START_PROCESSING(SSobj, src)
+	COOLDOWN_START(src, next_nag_time, bonus_countdown) //First nag after we've lost our speed bonus
+	radio.talk_into(src,"Created in [get_area(src)] by [bounty_holder] ([bounty_holder_job]). Speedy delivery bonus lost in [time2text(bonus_countdown,"mm:ss")].", RADIO_CHANNEL_SUPPLY)
+
+/obj/item/bounty_cube/test_cube
+	name = "test bounty cube"
+	desc = "It needs to be squeezed by someone with a bank account it can detect to set itself up. This will alert Supply."
```
[the remark this answers is not in the document]
Comment (cacogen) [reply code/game/machinery/civilian_bounties.dm:324]: I wasn't sure how IC to make it. I figured if it showed up in the round somehow then it's less immersion breaking this way. But sure.

**Comment by cacogen** [comment]
@coiax okay well whoever you're ready

**Comment by tgstation-server** [comment]
Your Fix/Feature pull request delta is currently below zero (-13). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.

**Pull request merged**